File: week-5/lecture/6-heap-sort/heap-sort.cpp
/**
 * @file heap-sort.cpp
 * @brief C++ version of the week 5 heap-sort lecture example.
 */

#include <iostream>
#include <vector>

using namespace std;

/**
 * @brief Restore the heap property for the subtree rooted at a given index.
 * @param arr vector<int>& List of integers arranged as a binary heap in an array.
 * @param n int Number of heap elements currently considered active.
 * @param i int Root index of the subtree that may violate the heap property.
 * @return void This function updates arr in place and does not return a value.
 */
void heapify(vector<int>& arr, int n, int i) {
    int largest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;

    if (left < n && arr[left] > arr[largest]) {
        largest = left;
    }
    if (right < n && arr[right] > arr[largest]) {
        largest = right;
    }
    if (largest != i) {
        swap(arr[i], arr[largest]);
        heapify(arr, n, largest);
    }
}

/**
 * @brief Sort the list in ascending order using heap sort.
 * @param arr vector<int>& List of integers to sort in place.
 * @return void This function updates arr in place and does not return a value.
 */
void heapSort(vector<int>& arr) {
    int n = static_cast<int>(arr.size());

    for (int i = n / 2; i >= 0; --i) {
        heapify(arr, n, i);
    }
    for (int i = n - 1; i > 0; --i) {
        swap(arr[0], arr[i]);
        heapify(arr, i, 0);
    }
}

/**
 * @brief Run the heap-sort lecture example.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    vector<int> arr = {1, 12, 9, 5, 6, 10};
    heapSort(arr);
    cout << "Sorted array is\n";
    for (int value : arr) {
        cout << value << ' ';
    }
    cout << '\n';
    return 0;
}

File: week-5/lecture/6-heap-sort/heap-sort.py
def heapify(arr, n, i):
    # Find the largest element among root and its children
    """Restore the heap property for the subtree rooted at the given index.
    
    Args:
        arr (list): The list being processed by the algorithm.
        n (int): The size of the heap.
        i (int): The index of the root node of the current subtree.
    
    Returns:
        None: This function modifies the input list in-place.
    """
    largest = i  # Assume root is the largest
    l = 2 * i + 1  # Left child index
    r = 2 * i + 2  # Right child index

    # Check if the left child is larger than the root
    if l < n and arr[l] > arr[largest]:
        largest = l

    # Check if the right child is larger than the current largest
    if r < n and arr[r] > arr[largest]:
        largest = r

    # If the largest is not the root, swap and continue heapifying
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]  # Swap
        heapify(arr, n, largest)  # Recursively heapify the affected subtree


def heapSort(arr):
    """Sort the list in ascending order using heap sort.
    
    Args:
        arr (list): The list to be sorted.
    
    Returns:
        None: This function modifies the input list in-place.
    """
    n = len(arr)

    # Build max heap
    for i in range(n//2 - 1, -1, -1):  # Start from the last non-leaf node
        heapify(arr, n, i)

    # Extract elements one by one from the heap and rebuild heap
    for i in range(n-1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]  # Swap the root (max element) with the last element

        # Restore the heap property for the reduced heap
        heapify(arr, i, 0)
  

arr = [1, 12, 9, 5, 6, 10]
heapSort(arr)
n = len(arr)
print("Sorted array is:")
for i in range(n):
    print("%d " % arr[i], end='')  # Print the sorted array